File: blockchain-mobile/lib/models/payment_history.dart
import 'enums/payment_status.dart';

class PaymentHistory {
  final int id;
  final String orderCode;
  final double amount;
  final String bankCode;
  final PaymentStatus paymentStatus;
  final DateTime payDate;
  final String reason;
  final int userInfoId;
  final double balance;

  PaymentHistory({
    required this.id,
    required this.orderCode,
    required this.amount,
    required this.bankCode,
    required this.paymentStatus,
    required this.payDate,
    required this.reason,
    required this.userInfoId,
    required this.balance,
  });

  factory PaymentHistory.fromJson(Map<String, dynamic> json) {
    return PaymentHistory(
      id: json['id'] as int,
      orderCode: json['orderCode'],
      amount: json['amount'] as double,
      bankCode: json['bankCode'] as String,
      paymentStatus: PaymentStatus.values.firstWhere(
          (element) => element.name == json["paymentStatus"],
          orElse: () => PaymentStatus.NULL),
      payDate: DateTime.parse(json['payDate'] as String),
      reason: json['reason'] as String,
      userInfoId: json['userInfoId'] as int,
      balance: (json['balance']?? 0),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'orderCode': orderCode,
      'amount': amount,
      'bankCode': bankCode,
      'paymentStatus': paymentStatus.name,
      'payDate': payDate.toIso8601String(),
      'reason': reason,
      'userInfoId': userInfoId,
      'balance': balance,
    };
  }

  @override
  String toString() {
    return 'PaymentHistory{id: $id, orderCode: $orderCode, amount: $amount, bankCode: $bankCode, paymentStatus: $paymentStatus, payDate: $payDate, reason: $reason, userInfoId: $userInfoId, balance: $balance}';
  }
}

File: blockchain-server/src/main/java/com/server/blockchainserver/models/user_model/PaymentHistory.java
package com.server.blockchainserver.models.user_model;

import com.server.blockchainserver.models.enums.PaymentStatus;
import jakarta.persistence.*;

import java.math.BigDecimal;
import java.time.Instant;

@Entity
@Table(name = "payment_history")
public class PaymentHistory {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;
    @Column(nullable = false)
    private String orderCode;
    @Column(nullable = false)
    private BigDecimal amount;
    @Column(nullable = false)
    private String bankCode;
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private PaymentStatus paymentStatus;
    @Column(nullable = false)
    private Instant payDate;

    @Column(nullable = true)
    private String reason;

    @ManyToOne
    @JoinColumn(name = "user_info_id", referencedColumnName = "id")
    private UserInfo userInfo;

    public PaymentHistory(String orderCode, BigDecimal amount, String bankCode,
                          PaymentStatus paymentStatus, Instant payDate, String reason, UserInfo userInfo) {
        this.orderCode = orderCode;
        this.amount = amount;
        this.bankCode = bankCode;
        this.paymentStatus = paymentStatus;
        this.payDate = payDate;
        this.reason = reason;
        this.userInfo = userInfo;
    }

    public PaymentHistory() {
    }

    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getOrderCode() {
        return orderCode;
    }

    public void setOrderCode(String orderCode) {
        this.orderCode = orderCode;
    }

    public BigDecimal getAmount() {
        return amount;
    }

    public void setAmount(BigDecimal amount) {
        this.amount = amount;
    }

    public String getBankCode() {
        return bankCode;
    }

    public void setBankCode(String bankCode) {
        this.bankCode = bankCode;
    }

    public PaymentStatus getPaymentStatus() {
        return paymentStatus;
    }

    public void setPaymentStatus(PaymentStatus paymentStatus) {
        this.paymentStatus = paymentStatus;
    }

    public Instant getPayDate() {
        return payDate;
    }

    public void setPayDate(Instant payDate) {
        this.payDate = payDate;
    }

    public String getReason() {
        return reason;
    }

    public void setReason(String reason) {
        this.reason = reason;
    }

    public UserInfo getUserInfo() {
        return userInfo;
    }

    public void setUserInfo(UserInfo userInfo) {
        this.userInfo = userInfo;
    }
}
